Adding descriptions for internal methods of LSP7 & LSP8 (#361)
lukso-network/docs: docs/standards/smart-contracts/lsp8-identifiable-digital-asset.md

:::

## Internal Functions

> :::info Warning

Reviewer: Same for the one above

#### Parameters:

|   Name   |   Type   | Description |
| :------- | :------- | :---------- |
| `from`   | address  | The sending address. |
| `to`     | address  | The receiving address. |
| `tokenId` | bytes32  | The token to transfer. |

:::note

#### Requirements:

- When `from` and `to` are both non-zero, ``from``'s `tokenId` will be transferred to `to`.
- When `from` is zero, `tokenId` will be minted for `to`.
- When `to` is zero, ``from``'s `tokenId` will be burned.

:::

### _notifyTokenSender

```
function _notifyTokenSender(
    address from,
    address to,
    bytes32 tokenId,
    bytes memory data
) internal virtual
```

An attempt is made to notify the token sender about the `tokenId` changing owners using LSP1 interface.

#### Parameters:

|   Name   |   Type   | Description |
| :------- | :------- | :---------- |
| `from`   | address  | The sending address. |
| `to`     | address  | The receiving address. |
| `tokenId` | bytes32  | The token to transfer. |
| `data`   | bytes    | Additional data the caller wants included in the emitted event, and sent in the hooks to `from` and `to` addresses. |

:::note

#### Requirements:

> -

Reviewer: empty Requirements, remove please

:::

### _notifyTokenReceiver

```
function _notifyTokenReceiver(
    address from,
    address to,
    bytes32 tokenId,
    bool force,
    bytes memory data
) internal virtual
```

An attempt is made to notify the token receiver about the `tokenId` changing owners using LSP1 interface. When force is FALSE the token receiver MUST support LSP1.

#### Parameters:

|   Name   |   Type   | Description |
| :------- | :------- | :---------- |
| `from`   | address  | The sending address. |
| `to`     | address  | The receiving address. |
| `tokenId` | bytes32  | The token to transfer. |
| `force`  | bool     | When set to TRUE, `to` may be any address; when set to FALSE `to` must be a contract that supports LSP1 UniversalReceiver and not revert. |
| `memory` | bytes    | Additional data the caller wants included in the emitted event, and sent in the hooks to `from` and `to` addresses. |

:::note

#### Requirements:

> - The receiver may revert when the token being sent is not wanted.

Reviewer: This is not a requ, lets remove it from here and add it to description

| `from`   | address  | The sending address. |
| `tokenId` | bytes32  | The token to burn. |
| `data`   | bytes    | Additional data the caller wants included in the emitted event, and sent in the hook to `from` address. |

:::note

#### Requirements:

- `from` cannot be the zero address.
- `tokenId` must exist.

:::

### _beforeTokenTransfer

```solidity
function _beforeTokenTransfer(
    address from,
    address to,
    bytes32 tokenId
) internal virtual
```

Hook that is called before any token transfer. This includes minting and burning.

#### Parameters:

|   Name   |   Type   | Description |
| :------- | :------- | :---------- |
| `from`   | address  | The sending address. |
| `to`     | address  | The receiving address. |
| `tokenId` | bytes32  | The token to transfer. |

:::note

#### Requirements:

- When `from` and `to` are both non-zero, ``from``'s `tokenId` will be transferred to `to`.
- When `from` is zero, `tokenId` will be minted for `to`.
- When `to` is zero, ``from``'s `tokenId` will be burned.

:::

### _notifyTokenSender

```solidity
function _notifyTokenSender(
    address from,
    address to,
    bytes32 tokenId,
    bytes memory data
) internal virtual
```

> An attempt is made to notify the token sender about the `tokenId` changing owners using LSP1 interface.

Reviewer: same above

| `data`   | bytes    | Additional data the caller wants included in the emitted event, and sent in the hooks to `from` and `to` addresses. |

### _notifyTokenReceiver

```solidity
function _notifyTokenReceiver(
    address from,
    address to,
    bytes32 tokenId,
    bool force,
    bytes memory data
) internal virtual
```

An attempt is made to notify the token receiver about the `tokenId` changing owners using LSP1 interface. When force is FALSE the token receiver MUST support LSP1.

#### Parameters:

|   Name   |   Type   | Description |
| :------- | :------- | :---------- |
| `from`   | address  | The sending address. |
| `to`     | address  | The receiving address. |
| `tokenId` | bytes32  | The token to transfer. |
| `force`  | bool     | When set to TRUE, `to` may be any address; when set to FALSE `to` must be a contract that supports LSP1 UniversalReceiver and not revert. |
| `memory` | bytes    | Additional data the caller wants included in the emitted event, and sent in the hooks to `from` and `to` addresses. |

### _revokeOperator

```solidity
function _revokeOperator(
    address operator,
    address tokenOwner,
    bytes32 tokenId
) internal virtual
```

> An attempt is made to notify the token receiver about the `tokenId` changing owners using LSP1 interface. When force is FALSE the token receiver MUST support LSP1.

Reviewer: not the right description

Author: No description in the natspec, I added custom description.

#### Parameters:

|     Name     |   Type   | Description |
| :----------- | :------- | :---------- |
| `operator`   | address | The address to revoke as an operator.     |
| `tokenOwner` | address | The address that is the owner of tokenId. |
| `tokenId`    | bytes32 | The token to disable operator status to.  |

### _clearOperators

> ```

Reviewer: please use ```solidity

Author: No description in the natspec, I added custom description.

function clearOperators(
    address tokenOwner,
    bytes32 tokenId
) internal virtual
```

> 

Reviewer: no description added

    address to,
    bytes32 tokenId,
    bool force,
    bytes memory data
) internal virtual
```

An attempt is made to notify the token receiver about the `tokenId` token being received by calling the **[universalReceiver(...)](./lsp0-erc725-account.md#universalreceiver)** function on the receiver address if it exists.

#### Parameters:

|   Name   |   Type   | Description |
| :------- | :------- | :---------- |
| `from`   | address  | The sending address. |
| `to`     | address  | The receiving address. |
| `tokenId` | bytes32  | The token to transfer. |
| `force`  | bool     | When set to TRUE, `to` may be any address; when set to FALSE `to` must be a contract that supports LSP1 UniversalReceiver and not revert. |
| `memory` | bytes    | Additional data the caller wants included in the emitted event, and sent in the hooks to `from` and `to` addresses. |

### _revokeOperator

```solidity
function _revokeOperator(
    address operator,
    address tokenOwner,
    bytes32 tokenId
) internal virtual
```

Revoke the `operator` of the `tokenId` token which belongs to `tokenOwner`.

#### Parameters:

|     Name     |   Type   | Description |
| :----------- | :------- | :---------- |

> | `operator`   | address | The address to revoke as an operator.     |

Reviewer 2: ```suggestion
| `operator`   | address | The address of the operator to revoke.     |
```

| Name       | Type    | Description                                                                                                             |
| :--------- | :------ | :---------------------------------------------------------------------------------------------------------------------- |

> | `operator` | address | The address of operator sender tokens.                                                                                 |

Reviewer: it seems strange, please revert the changes
